File: leetcode/easy/500-keyboardRow.js
// Given a List of words, return the words that can be typed using letters of alphabet on only one row's of American keyboard like the image below.
 
// Example:

// Input: ["Hello", "Alaska", "Dad", "Peace"]
// Output: ["Alaska", "Dad"]

// first pass solution using regex 
var findWords = function(words) {
    let top = /^[qwertyuiop]+$/;
    let middle = /^[asdfghjkl]+$/;
    let bottom = /^[zxcvbnm]+$/;
    let results = [];
    
    for (let word of words) {
        curr = word.toLowerCase();
        if (curr.match(top) || curr.match(middle) || curr.match(bottom)) {
            results.push(word)
        }
    }
    return results;
};

// second pass solution using sets and a nested loop to count chars
var findWords = function(words) {
    let top = new Set(["q", "w", "e", "r", "t", "y", "u", "i", "o", "p"]);
    let middle = new Set(["a", "s", "d", "f", "g", "h", "j", "k", "l"]);
    let bottom = new Set(["z", "x", "c", "v", "b", "n", "m"]);
    let results = [];
    
    for (let word of words) {
        curr = word.toLowerCase();
        let tLen = 0;
        let mLen = 0;
        let bLen = 0;
        
        for (let char of curr) {
            if (top.has(char)) tLen++;
            if (middle.has(char)) mLen++;
            if (bottom.has(char)) bLen++;
        }
        
        if (tLen === word.length || mLen === word.length || bLen === word.length) {
            results.push(word);
        }
    }
    return results;
};

File: leetcode/easy/500-keyboardRow.cpp
class Solution {
public:
    vector<string> findWords(vector<string>& words) {
        set<char> row1 = {'q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'};
        set<char> row2 = {'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l'};
        set<char> row3 = {'z', 'x', 'c', 'v', 'b', 'n', 'm'};
        vector<string> output;

        for (const auto word: words) {
            bool hasRow1 = false;
            bool hasRow2 = false;
            bool hasRow3 = false;
            int count = 0;
            
            for (int i = 0; i < word.length(); i++) {
                char c = tolower(word[i]);

                if (row1.find(c) != row1.end()) hasRow1 = true;
                if (row2.find(c) != row2.end()) hasRow2 = true;
                if (row3.find(c) != row3.end()) hasRow3 = true;
            }

            if (hasRow1) count++;
            if (hasRow2) count++;
            if (hasRow3) count++;

           if (count == 1) output.push_back(word);
        }
        return output;
    }
};
